src/topology/dense_embedding.lean

```diff
@@ -285,25 +289,51 @@ protected lemma subtype (p : α → Prop) : dense_embedding (subtype_emb p e) :=
 end dense_embedding
 
 lemma is_closed_property [topological_space β] {e : α → β} {p : β → Prop}
-  (he : closure (range e) = univ) (hp : is_closed {x | p x}) (h : ∀a, p (e a)) :
+  (he : dense_range e) (hp : is_closed {x | p x}) (h : ∀a, p (e a)) :
   ∀b, p b :=
 have univ ⊆ {b | p b},
-  from calc univ = closure (range e) : he.symm
+  from calc univ = closure (range e) : he.closure_range.symm
     ... ⊆ closure {b | p b} : closure_mono $ range_subset_iff.mpr h
     ... = _ : closure_eq_of_is_closed hp,
 assume b, this trivial
 
-lemma is_closed_property2 [topological_space α] [topological_space β] {e : α → β} {p : β → β → Prop}
-  (he : dense_embedding e) (hp : is_closed {q:β×β | p q.1 q.2}) (h : ∀a₁ a₂, p (e a₁) (e a₂)) :
+lemma is_closed_property2 [topological_space β] {e : α → β} {p : β → β → Prop}
+  (he : dense_range e) (hp : is_closed {q:β×β | p q.1 q.2}) (h : ∀a₁ a₂, p (e a₁) (e a₂)) :
   ∀b₁ b₂, p b₁ b₂ :=
 have ∀q:β×β, p q.1 q.2,
-  from is_closed_property (he.prod he).to_dense_inducing.closure_range hp $ assume a, h _ _,
+  from is_closed_property (he.prod he) hp $ λ _, h _ _,
 assume b₁ b₂, this ⟨b₁, b₂⟩
 
-lemma is_closed_property3 [topological_space α] [topological_space β] {e : α → β} {p : β → β → β → Prop}
-  (he : dense_embedding e) (hp : is_closed {q:β×β×β | p q.1 q.2.1 q.2.2}) (h : ∀a₁ a₂ a₃, p (e a₁) (e a₂) (e a₃)) :
+lemma is_closed_property3 [topological_space β] {e : α → β} {p : β → β → β → Prop}
+  (he : dense_range e) (hp : is_closed {q:β×β×β | p q.1 q.2.1 q.2.2}) (h : ∀a₁ a₂ a₃, p (e a₁) (e a₂) (e a₃)) :
   ∀b₁ b₂ b₃, p b₁ b₂ b₃ :=
 have ∀q:β×β×β, p q.1 q.2.1 q.2.2,
-  from is_closed_property (he.prod $ he.prod he).to_dense_inducing.closure_range hp $
-    assume ⟨a₁, a₂, a₃⟩, h _ _ _,
+  from is_closed_property (he.prod $ he.prod he) hp $ λ _, h _ _ _,
 assume b₁ b₂ b₃, this ⟨b₁, b₂, b₃⟩
+
+lemma is_closed_property' [topological_space β] {e : α → β} {p : β → Prop}
+  (he : closure (range e) = univ) (hp : is_closed {x | p x}) (h : ∀a, p (e a)) :
+  ∀b, p b := is_closed_property (dense_range_iff_closure_range.mpr he) hp h
+
+lemma is_closed_property2' [topological_space β] {e : α → β} {p : β → β → Prop}
+  (he : closure (range e) = univ) (hp : is_closed {q:β×β | p q.1 q.2}) (h : ∀a₁ a₂, p (e a₁) (e a₂)) :
+  ∀b₁ b₂, p b₁ b₂ := is_closed_property2 (dense_range_iff_closure_range.mpr he) hp h
+
+lemma is_closed_property3' [topological_space β] {e : α → β} {p : β → β → β → Prop}
+  (he : closure (range e) = univ) (hp : is_closed {q:β×β×β | p q.1 q.2.1 q.2.2}) (h : ∀a₁ a₂ a₃, p (e a₁) (e a₂) (e a₃)) :
+  ∀b₁ b₂ b₃, p b₁ b₂ b₃ := is_closed_property3 (dense_range_iff_closure_range.mpr he) hp h
```

Comment:
Do you really need these? I see that you use them below in `completionLean`, for instance in the proof of `induction_on`, written as `is_closed_property' dense hp ih a`. It made me wonder if the property `dense` you use here should not be reformulated in terms of `dense_range`, to have something coherent all over the way. What do you think?

Reply:
Yes, maybe I should be more authoritarian and ban those. I'll do it tomorrow if nobody complains until then.

Reply:
Done